File: example/step-definitions/screenshot.steps.ts
import { Then } from '@cucumber/cucumber';
import { ConductorWorld } from 'conductor';
import * as fs from 'fs';
import * as path from 'path';

const MAESTRO_TIMEOUT = { timeout: 120000 };

Then('I take a screenshot {string}', MAESTRO_TIMEOUT, async function (this: ConductorWorld, name: string) {
  const slug = name.replace(/\s+/g, '-').toLowerCase();
  const dir = 'reports/screenshots';
  fs.mkdirSync(dir, { recursive: true });

  if (this.web.isLaunched) {
    const screenshot = await this.web.takeScreenshot(slug);
    await this.attach(screenshot, 'image/png');
    this.logger.info({ path: `${dir}/${slug}.png` }, 'Web screenshot captured');
  }

  if (this.config.mobile.deviceId) {
    try {
      const screenshotPath = await this.maestro.takeScreenshot(slug);
      const buffer = fs.readFileSync(screenshotPath);
      await this.attach(buffer, 'image/png');
      this.logger.info({ path: screenshotPath }, 'Mobile screenshot captured');
    } catch {
      this.logger.warn('Mobile screenshot skipped — no device or Maestro unavailable');
    }
  }
});

Then('I take a web screenshot {string}', async function (this: ConductorWorld, name: string) {
  const slug = name.replace(/\s+/g, '-').toLowerCase();
  fs.mkdirSync('reports/screenshots', { recursive: true });
  const screenshot = await this.web.takeScreenshot(slug);
  await this.attach(screenshot, 'image/png');
});

Then('I take a mobile screenshot {string}', MAESTRO_TIMEOUT, async function (this: ConductorWorld, name: string) {
  const slug = name.replace(/\s+/g, '-').toLowerCase();
  fs.mkdirSync('reports/screenshots', { recursive: true });
  const screenshotPath = await this.maestro.takeScreenshot(slug);
  const buffer = fs.readFileSync(screenshotPath);
  await this.attach(buffer, 'image/png');
});

File: java/conductor-example/src/test/java/com/nouhouari/conductor/example/stepdefs/ScreenshotSteps.java
package com.nouhouari.conductor.example.stepdefs;

import com.nouhouari.conductor.world.ConductorWorld;
import io.cucumber.java.en.Then;

import java.io.IOException;
import java.io.UncheckedIOException;
import java.nio.file.Files;
import java.nio.file.Path;

/** Java port of example/step-definitions/screenshot.steps.ts — cross-cutting screenshot steps. */
public class ScreenshotSteps {

    private final ConductorWorld world;

    public ScreenshotSteps(ConductorWorld world) {
        this.world = world;
    }

    @Then("I take a screenshot {string}")
    public void iTakeAScreenshot(String name) {
        String slug = name.replaceAll("\\s+", "-").toLowerCase();
        ensureScreenshotsDir();

        if (world.web().isLaunched()) {
            byte[] screenshot = world.web().takeScreenshot(slug);
            world.scenario.attach(screenshot, "image/png", slug);
            world.logger.info("Web screenshot captured: reports/screenshots/{}.png", slug);
        }

        if (world.config.mobile().deviceId() != null) {
            try {
                byte[] buffer = world.maestro().takeScreenshot(slug);
                world.scenario.attach(buffer, "image/png", slug);
                world.logger.info("Mobile screenshot captured: {}", slug);
            } catch (RuntimeException e) {
                world.logger.warn("Mobile screenshot skipped — no device or Maestro unavailable");
            }
        }
    }

    @Then("I take a web screenshot {string}")
    public void iTakeAWebScreenshot(String name) {
        String slug = name.replaceAll("\\s+", "-").toLowerCase();
        ensureScreenshotsDir();
        byte[] screenshot = world.web().takeScreenshot(slug);
        world.scenario.attach(screenshot, "image/png", slug);
    }

    @Then("I take a mobile screenshot {string}")
    public void iTakeAMobileScreenshot(String name) {
        String slug = name.replaceAll("\\s+", "-").toLowerCase();
        ensureScreenshotsDir();
        byte[] buffer = world.maestro().takeScreenshot(slug);
        world.scenario.attach(buffer, "image/png", slug);
    }

    private static void ensureScreenshotsDir() {
        try {
            Files.createDirectories(Path.of("reports/screenshots"));
        } catch (IOException e) {
            throw new UncheckedIOException(e);
        }
    }
}
